presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -349,6 +378,49 @@ public synchronized boolean isClearTransactionId()
         return Optional.empty();
     }
 
+    private synchronized QueryResults getNextResultWithRetry(long token, UriInfo uriInfo, String scheme, DataSize targetResultSize)
```

Comment:
I think we'll want to add a configuration to specify what's the max duration of a query that is eligible for retry.

Given Presto's resource management, which is permissive after a query is dequeued from the resource groups, if we retry a query that runs for several hours and uses lots of resources, retries would multiply the resource usage of this query by the number of retries.  We should set some reasonable upper bound, maybe 5 minutes.